<!DOCTYPE html>
<html lang="en" itemscope itemtype="http://schema.org/WebPage">
<head><script src="/blog/livereload.js?mindelay=10&amp;v=2&amp;port=45315&amp;path=blog/livereload" data-no-instant defer></script>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>scRNAseq analysis reveals Immune Cell Signatures in Alzheimer’s Disease - Example R codes for BIOS 850 &amp; 880</title>
  

<meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=yes"/>

<meta name="MobileOptimized" content="width"/>
<meta name="HandheldFriendly" content="true"/>


<meta name="applicable-device" content="pc,mobile">

<meta name="theme-color" content="#f8f5ec" />
<meta name="msapplication-navbutton-color" content="#f8f5ec">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="#f8f5ec">

<meta name="mobile-web-app-capable" content="yes">

<meta name="author" content="LYQG" />
  <meta name="description" content="Single-Cell RNA Sequencing of Peripheral Blood Reveals Immune Cell Signatures in Alzheimer’s Disease
" />
<meta name="keywords" content="Alzheimer’s disease, Immune Cell Signatures, R, scRNAseq" />







<meta name="generator" content="Hugo 0.126.1" />


<link rel="canonical" href="http://localhost:45315/blog/post/immunead/" />





<link rel="icon" href="/blog/favicon.ico" />











<link rel="stylesheet" href="/blog/sass/jane.min.de22abc00de44766eebd1054fd9e0b254b071f89d5019044f893c484a9249a8d.css" integrity="sha256-3iKrwA3kR2buvRBU/Z4LJUsHH4nVAZBE&#43;JPEhKkkmo0=" media="screen" crossorigin="anonymous">







<meta property="og:url" content="http://localhost:45315/blog/post/immunead/">
  <meta property="og:site_name" content="Example R codes for BIOS 850 &amp; 880">
  <meta property="og:title" content="scRNAseq analysis reveals Immune Cell Signatures in Alzheimer’s Disease">
  <meta property="og:description" content="Single-Cell RNA Sequencing of Peripheral Blood Reveals Immune Cell Signatures in Alzheimer’s Disease">
  <meta property="og:locale" content="en">
  <meta property="og:type" content="article">
    <meta property="article:section" content="post">
    <meta property="article:published_time" content="2023-02-21T00:00:00+00:00">
    <meta property="article:modified_time" content="2023-02-21T00:00:00+00:00">
    <meta property="article:tag" content="Alzheimer’s Disease">
    <meta property="article:tag" content="Immune Cell Signatures">
    <meta property="article:tag" content="R">
    <meta property="article:tag" content="ScRNAseq">

  <meta itemprop="name" content="scRNAseq analysis reveals Immune Cell Signatures in Alzheimer’s Disease">
  <meta itemprop="description" content="Single-Cell RNA Sequencing of Peripheral Blood Reveals Immune Cell Signatures in Alzheimer’s Disease">
  <meta itemprop="datePublished" content="2023-02-21T00:00:00+00:00">
  <meta itemprop="dateModified" content="2023-02-21T00:00:00+00:00">
  <meta itemprop="wordCount" content="2317">
  <meta itemprop="keywords" content="Alzheimer’s disease,Immune Cell Signatures,R,scRNAseq">
  <meta name="twitter:card" content="summary">
  <meta name="twitter:title" content="scRNAseq analysis reveals Immune Cell Signatures in Alzheimer’s Disease">
  <meta name="twitter:description" content="Single-Cell RNA Sequencing of Peripheral Blood Reveals Immune Cell Signatures in Alzheimer’s Disease">

<!--[if lte IE 9]>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js"></script>
<![endif]-->

<!--[if lt IE 9]>
  <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
<![endif]-->




</head>
<body>
  <div id="mobile-navbar" class="mobile-navbar">
  <div class="mobile-header-logo">
    <a href="/" class="logo">Example R codes for BIOS 850 & 880</a>
  </div>
  <div class="mobile-navbar-icon">
    <span></span>
    <span></span>
    <span></span>
  </div>
</div>
<nav id="mobile-menu" class="mobile-menu slideout-menu">
  <ul class="mobile-menu-list">
    <li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/">Home</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/post/">Archives</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/tags/">Tags</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/categories/">Categories</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/about/">About</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/">external-link</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          <div class="mobile-menu-parent">
            <span class="mobile-submenu-open"></span>
            <a href="http://localhost:45315/blog/">
              docs
            </a>
          </div>
          <ul class="mobile-submenu-list">
            
              <li>
                <a href="http://localhost:45315/blog/post/ca/">Cluster analysis</a>
              </li>
            
              <li>
                <a href="http://localhost:45315/blog/post/knn/">K Nearest Neighbors</a>
              </li>
            
          </ul>
        
      </li>
    

    
  </ul>
</nav>


  
    






  <link rel="stylesheet" href="/lib/photoswipe/photoswipe.min.css" />
  <link rel="stylesheet" href="/lib/photoswipe/default-skin/default-skin.min.css" />




<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

<div class="pswp__bg"></div>

<div class="pswp__scroll-wrap">
    
    <div class="pswp__container">
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
    </div>
    
    <div class="pswp__ui pswp__ui--hidden">
    <div class="pswp__top-bar">
      
      <div class="pswp__counter"></div>
      <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
      <button class="pswp__button pswp__button--share" title="Share"></button>
      <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
      <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>
      
      
      <div class="pswp__preloader">
        <div class="pswp__preloader__icn">
          <div class="pswp__preloader__cut">
            <div class="pswp__preloader__donut"></div>
          </div>
        </div>
      </div>
    </div>
    <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
      <div class="pswp__share-tooltip"></div>
    </div>
    <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
    </button>
    <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
    </button>
    <div class="pswp__caption">
      <div class="pswp__caption__center"></div>
    </div>
    </div>
    </div>
</div>

  

  

  

  <header id="header" class="header container">
    <div class="logo-wrapper">
  <a href="/" class="logo">
    
      Example R codes for BIOS 850 & 880
    
  </a>
</div>

<nav class="site-navbar">
  <ul id="menu" class="menu">
    
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/">Home</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/post/">Archives</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/tags/">Tags</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/categories/">Categories</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/about/">About</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="http://localhost:45315/blog/">external-link</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          <a class="menu-item-link menu-parent" href="http://localhost:45315/blog/">docs</a>
          <ul class="submenu">
            
              <li class="submenu-item">
                <a href="http://localhost:45315/blog/post/ca/">Cluster analysis</a>
              </li>
            
              <li class="submenu-item">
                <a href="http://localhost:45315/blog/post/knn/">K Nearest Neighbors</a>
              </li>
            
          </ul>

        

      </li>
    

    
    

    
  </ul>
</nav>

  </header>

  <div id="mobile-panel">
    <main id="main" class="main bg-llight wallpaper">
      <div class="content-wrapper">
        <div id="content" class="content container">
          <article class="post bg-white">
    
    <header class="post-header">
      <h1 class="post-title">scRNAseq analysis reveals Immune Cell Signatures in Alzheimer’s Disease</h1>
      

      <div class="post-meta">
  <div class="post-meta-author">
    by
      LYQG
    
  </div>

  <div class="post-meta-time">
    <time datetime="2023-02-21">
      2023-02-21
    </time>
  </div>

  


  <div class="post-meta__right">
    

    <div class="post-meta-category">
        <a href="http://localhost:45315/blog/categories/alzheimers-disease/"> Alzheimer’s disease </a>
          <a href="http://localhost:45315/blog/categories/immune-cell-signatures/"> Immune Cell Signatures </a>
          <a href="http://localhost:45315/blog/categories/r/"> R </a>
          <a href="http://localhost:45315/blog/categories/scrnaseq/"> scRNAseq </a>
          
      </div>


    
    


    
    
  </div>
</div>

    </header>

    
    


    
    <div class="post-content">
      <p>Single-Cell RNA Sequencing of Peripheral Blood Reveals Immune Cell Signatures in Alzheimer’s Disease</p>
<p><img alt="paper_header" src="/blog/post/immunead/post/immunead/paper.png">
<img alt="paper_figure" src="/blog/post/immunead/post/immunead/Fig1.png"></p>
<p>GEO data: GSE181279</p>
<p>Article: Single-Cell RNA Sequencing of Peripheral Blood Reveals Immune Cell Signatures in Alzheimer’s Disease</p>
<p>Journal: frontiers in Immunology</p>
<p>Abstract: The peripheral immune system is thought to affect the pathology of the central nervous
system in Alzheimer’s disease (AD). However, current knowledge is inadequate for
understanding the characteristics of peripheral immune cells in AD. This study aimed to
explore the molecular basis of peripheral immune cells and the features of adaptive
immune repertoire at a single cell level. We proﬁled 36,849 peripheral blood mononuclear
cells from AD patients with amyloid-positive status and normal controls with amyloid-
negative status by 5’ single-cell transcriptome and immune repertoire sequencing using
the cell ranger standard analysis procedure. We revealed ﬁve immune cell subsets: CD4+
T cells, CD8+ T cells, B cells, natural killer cells, and monocytes–macrophages cells, and
disentangled the characteristic alterations of cell subset proportion and gene expression
patterns in AD. Thirty-one cell type-speciﬁc key genes, comprising abundant human
leukocyte antigen genes, and multiple immune-related pathways were identiﬁed by
protein–protein interaction network and pathway enrichment analysis. We also found
high-frequency ampliﬁcation clonotypes in T and B cells and decreased diversity in T cells
in AD. As clone ampliﬁcation suggested the activation of an adaptive immune response
against speciﬁc antigens, we speculated that the peripheral adaptive immune response,
especially mediated by T cells, may have a role in the pathogenesis of AD. This ﬁnding may
also contribute to further research regarding disease mechanism and the development of
immune-related biomarkers or therapy.</p>
<p>摘要：外周免疫系统被认为影响阿尔茨海默病（AD）的中枢神经病理。然而，目前的知识还不足以了解AD患者的外周免疫细胞的特征。本研究旨在探索外周免疫细胞的分子基础和适应性免疫库的特征。我们通过5‘单细胞转录组和免疫库测序，分析了36849个外周血单核细胞(来自淀粉样蛋白阳性AD患者和阴性患者)。我们发现了5个免疫细胞亚群： CD4+ T细胞、CD8+ T细胞、B细胞、自然杀伤细胞和单核-巨噬细胞，并揭示了AD中细胞亚群比例和基因表达模式的特征性改变。通过蛋白-蛋白相互作用网络和通路富集分析，鉴定出31个细胞类型特异性的关键基因，其中包括丰富的人类白细胞抗原基因和多个免疫相关通路。我们还发现在T和B细胞中的高频扩增克隆子，在AD中T细胞多样性降低。由于克隆扩增表明激活了针对特定抗原的适应性免疫反应，我们推测外周适应性免疫反应更可能由T细胞介导的，可能在AD的发病机制中起作用。这一发现有助于进一步研究疾病机制和免疫相关生物标志物或治疗的发展。</p>
<h2 id="step-1-导入数据">Step 1: 导入数据</h2>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#a6e22e">setwd</span>(<span style="color:#e6db74">&#34;/media/lyqg/WD_BLACK/000000_YanmingLi_Work_Z820/000000复现/AD疾病的PBMC单细胞转录组实战&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">###### step1:导入数据 ######   </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">## 解压缩到 GSE181279_raw/， 名字如“GSM5494107_AD1_barcodes.tsv.gz”， 改为“GSM5494107-AD1_barcodes.tsv.gz”</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">library</span>(stringr)
</span></span><span style="display:flex;"><span>fs <span style="color:#f92672">=</span> <span style="color:#a6e22e">list.files</span>(<span style="color:#e6db74">&#39;./GSE181279_raw/&#39;</span>,pattern <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;^GSM&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#75715e">#执行这一步需要解压tar -xvf</span>
</span></span><span style="display:flex;"><span>samples<span style="color:#f92672">=</span><span style="color:#a6e22e">str_split</span>(fs,<span style="color:#e6db74">&#39;_&#39;</span>,simplify <span style="color:#f92672">=</span> T)[,<span style="color:#ae81ff">1</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">lapply</span>(<span style="color:#a6e22e">unique</span>(samples),<span style="color:#66d9ef">function</span>(x){
</span></span><span style="display:flex;"><span>  y<span style="color:#f92672">=</span>fs<span style="color:#a6e22e">[grepl</span>(x,fs)]
</span></span><span style="display:flex;"><span>  folder<span style="color:#f92672">=</span><span style="color:#a6e22e">paste0</span>(<span style="color:#e6db74">&#34;GSE181279_raw/&#34;</span>, <span style="color:#a6e22e">str_split</span>(y[1],<span style="color:#e6db74">&#39;_&#39;</span>,simplify <span style="color:#f92672">=</span> T)[,<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">dir.create</span>(folder,recursive <span style="color:#f92672">=</span> T)
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#为每个样本创建子文件夹</span>
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">file.rename</span>(<span style="color:#a6e22e">paste0</span>(<span style="color:#e6db74">&#34;GSE181279_raw/&#34;</span>,y[1]),<span style="color:#a6e22e">file.path</span>(folder,<span style="color:#e6db74">&#34;barcodes.tsv.gz&#34;</span>))
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#重命名文件，并移动到相应的子文件夹里</span>
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">file.rename</span>(<span style="color:#a6e22e">paste0</span>(<span style="color:#e6db74">&#34;GSE181279_raw/&#34;</span>,y[2]),<span style="color:#a6e22e">file.path</span>(folder,<span style="color:#e6db74">&#34;features.tsv.gz&#34;</span>))
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">file.rename</span>(<span style="color:#a6e22e">paste0</span>(<span style="color:#e6db74">&#34;GSE181279_raw/&#34;</span>,y[3]),<span style="color:#a6e22e">file.path</span>(folder,<span style="color:#e6db74">&#34;matrix.mtx.gz&#34;</span>))
</span></span><span style="display:flex;"><span>})
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dir<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;./GSE181279_raw/&#39;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>samples<span style="color:#f92672">=</span><span style="color:#a6e22e">list.files</span>( dir )
</span></span><span style="display:flex;"><span>samples
</span></span><span style="display:flex;"><span><span style="color:#75715e"># samples = head(samples,10) </span>
</span></span><span style="display:flex;"><span>sceList <span style="color:#f92672">=</span> <span style="color:#a6e22e">lapply</span>(samples,<span style="color:#66d9ef">function</span>(pro){ 
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># pro=samples[1] </span>
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">print</span>(pro)  
</span></span><span style="display:flex;"><span>  sce <span style="color:#f92672">=</span><span style="color:#a6e22e">CreateSeuratObject</span>(counts <span style="color:#f92672">=</span>  <span style="color:#a6e22e">Read10X</span>(<span style="color:#a6e22e">file.path</span>(dir,pro )) ,
</span></span><span style="display:flex;"><span>                          project <span style="color:#f92672">=</span>  <span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#39;^GSM[0-9]*_&#39;</span>,<span style="color:#e6db74">&#39;&#39;</span>,pro)  ,
</span></span><span style="display:flex;"><span>                          min.cells <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>,
</span></span><span style="display:flex;"><span>                          min.features <span style="color:#f92672">=</span> <span style="color:#ae81ff">500</span> )
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">return</span>(sce)
</span></span><span style="display:flex;"><span>})
</span></span><span style="display:flex;"><span><span style="color:#75715e">#gsub函数： gsub (匹配内容，替换内容，操作对象）</span>
</span></span><span style="display:flex;"><span>samples
</span></span><span style="display:flex;"><span>a<span style="color:#f92672">=</span><span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#39;^GSM[0-9]*_&#39;</span>,<span style="color:#e6db74">&#39;&#39;</span>,samples)
</span></span><span style="display:flex;"><span>a
</span></span><span style="display:flex;"><span>b<span style="color:#f92672">=</span><span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#39;_gene_cell_exprs_table.txt.gz&#39;</span>,<span style="color:#e6db74">&#39;&#39;</span>,<span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#39;^GSM[0-9]*_&#39;</span>,<span style="color:#e6db74">&#39;&#39;</span>,samples))
</span></span><span style="display:flex;"><span>b
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">names</span>(sceList) 
</span></span><span style="display:flex;"><span><span style="color:#75715e"># gsub(&#39;^GSM[0-9]*&#39;,&#39;&#39;,samples)</span>
</span></span><span style="display:flex;"><span>sce.all<span style="color:#f92672">=</span><span style="color:#a6e22e">merge</span>(x<span style="color:#f92672">=</span>sceList[[1]],
</span></span><span style="display:flex;"><span>              y<span style="color:#f92672">=</span>sceList[ <span style="color:#ae81ff">-1</span> ],
</span></span><span style="display:flex;"><span>              add.cell.ids <span style="color:#f92672">=</span>  <span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#39;_gene_cell_exprs_table.txt.gz&#39;</span>,<span style="color:#e6db74">&#39;&#39;</span>,<span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#39;^GSM[0-9]*_&#39;</span>,<span style="color:#e6db74">&#39;&#39;</span>,samples) )     )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">as.data.frame</span>(sce.all<span style="color:#f92672">@</span>assays<span style="color:#f92672">$</span>RNA<span style="color:#f92672">@</span>counts[1<span style="color:#f92672">:</span><span style="color:#ae81ff">10</span>, <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#ae81ff">2</span>])
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">head</span>(sce.all<span style="color:#f92672">@</span>meta.data, <span style="color:#ae81ff">10</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">table</span>(sce.all<span style="color:#f92672">$</span>orig.ident) 
</span></span></code></pre></div><p><img alt="screen_shot 1" src="/blog/post/immunead/post/immunead/screenshot1.png"></p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#75715e"># 分组信息，通常是隐含在文件名，样品名字里面</span>
</span></span><span style="display:flex;"><span>phe<span style="color:#f92672">=</span><span style="color:#a6e22e">str_split</span>( <span style="color:#a6e22e">colnames</span>(sce.all),<span style="color:#e6db74">&#39;[-_]&#39;</span>,simplify <span style="color:#f92672">=</span> T)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">head</span>(phe)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">table</span>(phe[,<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span><span style="color:#75715e">#sce.all$group=toupper( substring(colnames(sce.all),1,2))</span>
</span></span><span style="display:flex;"><span>sce.all<span style="color:#f92672">$</span>group<span style="color:#f92672">=</span><span style="color:#a6e22e">toupper</span>( <span style="color:#a6e22e">substring</span>(phe[,<span style="color:#ae81ff">2</span>],<span style="color:#ae81ff">1</span>,<span style="color:#ae81ff">2</span>))
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">table</span>(sce.all<span style="color:#f92672">$</span>group)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">#substring函数：SUBSTRING(name,1,2) 截取name这个字段 从第1个字符开始 只截取之后的2个字符</span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">#toupper函数：转换为大写字母</span>
</span></span><span style="display:flex;"><span><span style="color:#75715e"># sce.all@meta.data$group =  ifelse(grepl(&#39;ne&#39;,sce.all$orig.ident),&#39;ne&#39;,&#39;e&#39;)</span>
</span></span><span style="display:flex;"><span><span style="color:#75715e"># sce.all$group=  gsub( &#39;^[0-9]*_&#39; ,&#39;&#39;,sce.all$orig.ident) </span>
</span></span><span style="display:flex;"><span> 
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">table</span>(sce.all<span style="color:#f92672">@</span>meta.data<span style="color:#f92672">$</span>group)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">table</span>(sce.all<span style="color:#f92672">@</span>meta.data<span style="color:#f92672">$</span>orig.ident)
</span></span></code></pre></div><p><img alt="screen_shot 2" src="/blog/post/immunead/post/immunead/screenshot2.png"></p>
<h2 id="step2-qc质控">step2: QC质控</h2>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#75715e">###### step2:QC质控 ######</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dir.create</span>(<span style="color:#e6db74">&#34;./1-QC&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">setwd</span>(<span style="color:#e6db74">&#34;./1-QC&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>input_sce<span style="color:#f92672">&lt;-</span>sce.all
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#计算线粒体基因比例</span>
</span></span><span style="display:flex;"><span>  mito_genes<span style="color:#f92672">=</span><span style="color:#a6e22e">rownames</span>(input_sce)<span style="color:#a6e22e">[grep</span>(<span style="color:#e6db74">&#34;^MT-&#34;</span>, <span style="color:#a6e22e">rownames</span>(input_sce),ignore.case <span style="color:#f92672">=</span> T)] 
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">print</span>(mito_genes) <span style="color:#75715e">#可能是13个线粒体基因</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#input_sce=PercentageFeatureSet(input_sce, &#34;^MT-&#34;, col.name = &#34;percent_mito&#34;)</span>
</span></span><span style="display:flex;"><span>  input_sce<span style="color:#f92672">=</span><span style="color:#a6e22e">PercentageFeatureSet</span>(input_sce, features <span style="color:#f92672">=</span> mito_genes, col.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;percent_mito&#34;</span>)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">fivenum</span>(input_sce<span style="color:#f92672">@</span>meta.data<span style="color:#f92672">$</span>percent_mito)<span style="color:#75715e">#fivenum函数：返回五个数据：最小值、下四分位数、中位数、上四分位数、最大值。</span>
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#计算核糖体基因比例</span>
</span></span><span style="display:flex;"><span>  ribo_genes<span style="color:#f92672">=</span><span style="color:#a6e22e">rownames</span>(input_sce)<span style="color:#a6e22e">[grep</span>(<span style="color:#e6db74">&#34;^Rp[sl]&#34;</span>, <span style="color:#a6e22e">rownames</span>(input_sce),ignore.case <span style="color:#f92672">=</span> T)]
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">print</span>(ribo_genes)<span style="color:#75715e">#可能是100个核糖体基因</span>
</span></span><span style="display:flex;"><span>  input_sce<span style="color:#f92672">=</span><span style="color:#a6e22e">PercentageFeatureSet</span>(input_sce,  features <span style="color:#f92672">=</span> ribo_genes, col.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;percent_ribo&#34;</span>)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">fivenum</span>(input_sce<span style="color:#f92672">@</span>meta.data<span style="color:#f92672">$</span>percent_ribo)
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#计算红血细胞基因比例</span>
</span></span><span style="display:flex;"><span>  Hb_genes<span style="color:#f92672">=</span><span style="color:#a6e22e">rownames</span>(input_sce)<span style="color:#a6e22e">[grep</span>(<span style="color:#e6db74">&#34;^Hb[^(p)]&#34;</span>, <span style="color:#a6e22e">rownames</span>(input_sce),ignore.case <span style="color:#f92672">=</span> T)]
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">print</span>(Hb_genes<span style="color:#75715e">#可能是7个红血细胞基因</span>
</span></span><span style="display:flex;"><span>  input_sce<span style="color:#f92672">=</span><span style="color:#a6e22e">PercentageFeatureSet</span>(input_sce,  features <span style="color:#f92672">=</span> Hb_genes,col.name <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;percent_hb&#34;</span>)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">fivenum</span>(input_sce<span style="color:#f92672">@</span>meta.data<span style="color:#f92672">$</span>percent_hb)
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#可视化细胞的上述比例情况</span>
</span></span><span style="display:flex;"><span>  feats <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;nFeature_RNA&#34;</span>, <span style="color:#e6db74">&#34;nCount_RNA&#34;</span>, <span style="color:#e6db74">&#34;percent_mito&#34;</span>, <span style="color:#e6db74">&#34;percent_ribo&#34;</span>, <span style="color:#e6db74">&#34;percent_hb&#34;</span>)
</span></span><span style="display:flex;"><span>  feats <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;nFeature_RNA&#34;</span>, <span style="color:#e6db74">&#34;nCount_RNA&#34;</span>)
</span></span><span style="display:flex;"><span>  p1<span style="color:#f92672">=</span><span style="color:#a6e22e">VlnPlot</span>(input_sce, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;orig.ident&#34;</span>, features <span style="color:#f92672">=</span> feats, pt.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>, ncol <span style="color:#f92672">=</span> <span style="color:#ae81ff">2</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">NoLegend</span>()
</span></span><span style="display:flex;"><span>  p1 
</span></span><span style="display:flex;"><span>  w<span style="color:#f92672">=</span><span style="color:#a6e22e">length</span>(<span style="color:#a6e22e">unique</span>(input_sce<span style="color:#f92672">$</span>orig.ident))<span style="color:#f92672">/</span><span style="color:#ae81ff">3+5</span>;w
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;Vlnplot1.pdf&#34;</span>,plot<span style="color:#f92672">=</span>p1,width <span style="color:#f92672">=</span> w,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>)
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  feats <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;percent_mito&#34;</span>, <span style="color:#e6db74">&#34;percent_ribo&#34;</span>, <span style="color:#e6db74">&#34;percent_hb&#34;</span>)
</span></span><span style="display:flex;"><span>  p2<span style="color:#f92672">=</span><span style="color:#a6e22e">VlnPlot</span>(input_sce, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;orig.ident&#34;</span>, features <span style="color:#f92672">=</span> feats, pt.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>, ncol <span style="color:#f92672">=</span> <span style="color:#ae81ff">3</span>, same.y.lims<span style="color:#f92672">=</span>T) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">scale_y_continuous</span>(breaks<span style="color:#f92672">=</span><span style="color:#a6e22e">seq</span>(<span style="color:#ae81ff">0</span>, <span style="color:#ae81ff">100</span>, <span style="color:#ae81ff">5</span>)) <span style="color:#f92672">+</span>
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">NoLegend</span>()
</span></span><span style="display:flex;"><span>  p2  
</span></span><span style="display:flex;"><span>  w<span style="color:#f92672">=</span><span style="color:#a6e22e">length</span>(<span style="color:#a6e22e">unique</span>(input_sce<span style="color:#f92672">$</span>orig.ident))<span style="color:#f92672">/</span><span style="color:#ae81ff">2+5</span>;w
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;Vlnplot2.pdf&#34;</span>,plot<span style="color:#f92672">=</span>p2,width <span style="color:#f92672">=</span> w,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>)
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  p3<span style="color:#f92672">=</span><span style="color:#a6e22e">FeatureScatter</span>(input_sce, <span style="color:#e6db74">&#34;nCount_RNA&#34;</span>, <span style="color:#e6db74">&#34;nFeature_RNA&#34;</span>, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;orig.ident&#34;</span>, pt.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.5</span>)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;Scatterplot.pdf&#34;</span>,plot<span style="color:#f92672">=</span>p3)
</span></span></code></pre></div><p><img alt="Vlnplot1" src="/blog/post/immunead/post/immunead/Vlnplot1.png">
<img alt="Vlnplot2" src="/blog/post/immunead/post/immunead/Vlnplot2.png">
<img alt="Scatterplot1" src="/blog/post/immunead/post/immunead/Scatterplot.png"></p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#75715e">#根据上述指标，过滤低质量细胞/基因</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#过滤指标1:最少表达基因数的细胞&amp;最少表达细胞数的基因</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># 一般来说，在CreateSeuratObject的时候已经是进行了这个过滤操作</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># 如果后期看到了自己的单细胞降维聚类分群结果很诡异，就可以回过头来看质量控制环节</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># 先走默认流程即可</span>
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">if</span>(F){
</span></span><span style="display:flex;"><span>    selected_c <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">WhichCells</span>(input_sce, expression <span style="color:#f92672">=</span> nFeature_RNA <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">500</span>)
</span></span><span style="display:flex;"><span>    selected_f <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">rownames</span>(input_sce)[Matrix<span style="color:#f92672">::</span><span style="color:#a6e22e">rowSums</span>(input_sce<span style="color:#f92672">@</span>assays<span style="color:#f92672">$</span>RNA<span style="color:#f92672">@</span>counts <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">0</span> ) <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">3</span>]
</span></span><span style="display:flex;"><span>    input_sce.filt <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">subset</span>(input_sce, features <span style="color:#f92672">=</span> selected_f, cells <span style="color:#f92672">=</span> selected_c)
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">dim</span>(input_sce) 
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">dim</span>(input_sce.filt) 
</span></span><span style="display:flex;"><span>  }
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  input_sce.filt <span style="color:#f92672">=</span>  input_sce
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#75715e"># par(mar = c(4, 8, 2, 1))</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e"># 这里的C 这个矩阵，有一点大，可以考虑随抽样 </span>
</span></span><span style="display:flex;"><span>  C<span style="color:#f92672">=</span><span style="color:#a6e22e">subset</span>(input_sce.filt,downsample<span style="color:#f92672">=</span><span style="color:#ae81ff">100</span>)<span style="color:#f92672">@</span>assays<span style="color:#f92672">$</span>RNA<span style="color:#f92672">@</span>counts <span style="color:#75715e"># downSample将随机抽样一个数据集</span>
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">dim</span>(C)
</span></span><span style="display:flex;"><span>  C<span style="color:#f92672">=</span>Matrix<span style="color:#f92672">::</span><span style="color:#a6e22e">t</span>(Matrix<span style="color:#f92672">::</span><span style="color:#a6e22e">t</span>(C)<span style="color:#f92672">/</span>Matrix<span style="color:#f92672">::</span><span style="color:#a6e22e">colSums</span>(C)) <span style="color:#f92672">*</span> <span style="color:#ae81ff">100</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>  most_expressed <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">order</span>(<span style="color:#a6e22e">apply</span>(C, <span style="color:#ae81ff">1</span>, median), decreasing <span style="color:#f92672">=</span> T)[50<span style="color:#f92672">:</span><span style="color:#ae81ff">1</span>]
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">pdf</span>(<span style="color:#e6db74">&#34;TOP50_most_expressed_gene.pdf&#34;</span>,width<span style="color:#f92672">=</span><span style="color:#ae81ff">21</span>,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">21</span>)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">boxplot</span>(<span style="color:#a6e22e">as.matrix</span>(Matrix<span style="color:#f92672">::</span><span style="color:#a6e22e">t</span>(C[most_expressed, ])),
</span></span><span style="display:flex;"><span>          cex <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.1</span>, las <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>, 
</span></span><span style="display:flex;"><span>          xlab <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;% total count per cell&#34;</span>, 
</span></span><span style="display:flex;"><span>          col <span style="color:#f92672">=</span> (scales<span style="color:#f92672">::</span><span style="color:#a6e22e">hue_pal</span>())(<span style="color:#ae81ff">50</span>)[50<span style="color:#f92672">:</span><span style="color:#ae81ff">1</span>], 
</span></span><span style="display:flex;"><span>          horizontal <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">dev.off</span>()
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">rm</span>(C)
</span></span></code></pre></div><p><img alt="Boxplot" src="/blog/post/immunead/post/immunead/TOP50_most_expressed_gene.png"></p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#75715e">#过滤指标2:线粒体/核糖体基因比例(根据上面的violin图)</span>
</span></span><span style="display:flex;"><span>  selected_mito <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">WhichCells</span>(input_sce.filt, expression <span style="color:#f92672">=</span> percent_mito <span style="color:#f92672">&lt;</span> <span style="color:#ae81ff">15</span>)
</span></span><span style="display:flex;"><span>  selected_ribo <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">WhichCells</span>(input_sce.filt, expression <span style="color:#f92672">=</span> percent_ribo <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">3</span>)
</span></span><span style="display:flex;"><span>  selected_hb <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">WhichCells</span>(input_sce.filt, expression <span style="color:#f92672">=</span> percent_hb <span style="color:#f92672">&lt;</span> <span style="color:#ae81ff">1</span> )
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">length</span>(selected_hb)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">length</span>(selected_ribo)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">length</span>(selected_mito)
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  input_sce.filt <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">subset</span>(input_sce.filt, cells <span style="color:#f92672">=</span> selected_mito)
</span></span><span style="display:flex;"><span>  input_sce.filt <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">subset</span>(input_sce.filt, cells <span style="color:#f92672">=</span> selected_ribo)
</span></span><span style="display:flex;"><span>  input_sce.filt <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">subset</span>(input_sce.filt, cells <span style="color:#f92672">=</span> selected_hb)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">dim</span>(input_sce.filt)<span style="color:#75715e">#16382 36166</span>
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">dim</span>(input_sce)<span style="color:#75715e">#16382 36252</span>
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">table</span>(input_sce.filt<span style="color:#f92672">$</span>orig.ident) 
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#75715e">#可视化过滤后的情况</span>
</span></span><span style="display:flex;"><span>  feats <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;nFeature_RNA&#34;</span>, <span style="color:#e6db74">&#34;nCount_RNA&#34;</span>)
</span></span><span style="display:flex;"><span>  p1_filtered<span style="color:#f92672">=</span><span style="color:#a6e22e">VlnPlot</span>(input_sce.filt, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;orig.ident&#34;</span>, features <span style="color:#f92672">=</span> feats, pt.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>, ncol <span style="color:#f92672">=</span> <span style="color:#ae81ff">2</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">NoLegend</span>()
</span></span><span style="display:flex;"><span>  p1
</span></span><span style="display:flex;"><span>  w<span style="color:#f92672">=</span><span style="color:#a6e22e">length</span>(<span style="color:#a6e22e">unique</span>(input_sce.filt<span style="color:#f92672">$</span>orig.ident))<span style="color:#f92672">/</span><span style="color:#ae81ff">3+5</span>;w 
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;Vlnplot1_filtered.pdf&#34;</span>,plot<span style="color:#f92672">=</span>p1_filtered,width <span style="color:#f92672">=</span> w,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>)
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  feats <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;percent_mito&#34;</span>, <span style="color:#e6db74">&#34;percent_ribo&#34;</span>, <span style="color:#e6db74">&#34;percent_hb&#34;</span>)
</span></span><span style="display:flex;"><span>  p2_filtered<span style="color:#f92672">=</span><span style="color:#a6e22e">VlnPlot</span>(input_sce.filt, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;orig.ident&#34;</span>, features <span style="color:#f92672">=</span> feats, pt.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>, ncol <span style="color:#f92672">=</span> <span style="color:#ae81ff">3</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>    <span style="color:#a6e22e">NoLegend</span>()
</span></span><span style="display:flex;"><span>  p2
</span></span><span style="display:flex;"><span>  w<span style="color:#f92672">=</span><span style="color:#a6e22e">length</span>(<span style="color:#a6e22e">unique</span>(input_sce.filt<span style="color:#f92672">$</span>orig.ident))<span style="color:#f92672">/</span><span style="color:#ae81ff">2+5</span>;w 
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;Vlnplot2_filtered.pdf&#34;</span>,plot<span style="color:#f92672">=</span>p2_filtered,width <span style="color:#f92672">=</span> w,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>) 
</span></span></code></pre></div><p><img alt="VlnFilterplot1" src="/blog/post/immunead/post/immunead/Vlnplot1_filtered.png">
<img alt="VlnFilterplot2" src="/blog/post/immunead/post/immunead/Vlnplot2_filtered.png"></p>
<p>print(dim(sce.all))
print(dim(input_sce.filt))
setwd(&rsquo;../')</p>
<p><img alt="dim" src="/blog/post/immunead/post/immunead/dim.png"></p>
<h2 id="step3-harmony整合多个单细胞样品">step3: harmony整合多个单细胞样品</h2>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#75715e">###### step3: harmony整合多个单细胞样品 ######</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dir.create</span>(<span style="color:#e6db74">&#34;2-harmony&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">setwd</span>(<span style="color:#e6db74">&#34;2-harmony&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">getwd</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e"># 默认 ScaleData 没有添加&#34;nCount_RNA&#34;, &#34;nFeature_RNA&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>input_sce<span style="color:#f92672">&lt;-</span>input_sce.filt
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">print</span>(<span style="color:#a6e22e">dim</span>(input_sce))
</span></span><span style="display:flex;"><span>  input_sce <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">NormalizeData</span>(input_sce, 
</span></span><span style="display:flex;"><span>                             normalization.method <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;LogNormalize&#34;</span>,
</span></span><span style="display:flex;"><span>                             scale.factor <span style="color:#f92672">=</span> <span style="color:#ae81ff">1e4</span>) 
</span></span><span style="display:flex;"><span>  input_sce <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">FindVariableFeatures</span>(input_sce)
</span></span><span style="display:flex;"><span>  input_sce <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">ScaleData</span>(input_sce)
</span></span><span style="display:flex;"><span>  input_sce <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">RunPCA</span>(input_sce, features <span style="color:#f92672">=</span> <span style="color:#a6e22e">VariableFeatures</span>(object <span style="color:#f92672">=</span> input_sce))
</span></span><span style="display:flex;"><span>  seuratObj <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">RunHarmony</span>(input_sce, <span style="color:#e6db74">&#34;orig.ident&#34;</span>)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">names</span>(seuratObj<span style="color:#f92672">@</span>reductions)
</span></span><span style="display:flex;"><span>  seuratObj <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">RunUMAP</span>(seuratObj,  dims <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#ae81ff">15</span>, 
</span></span><span style="display:flex;"><span>                       reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;harmony&#34;</span>) <span style="color:#75715e">#umap</span>
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#seuratObj=RunTSNE(seuratObj,  dims = 1:15, reduction = &#34;harmony&#34;) #tsne</span>
</span></span><span style="display:flex;"><span>  p <span style="color:#f92672">=</span> <span style="color:#a6e22e">DimPlot</span>(seuratObj,reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;umap&#34;</span>,label<span style="color:#f92672">=</span>T ) 
</span></span><span style="display:flex;"><span>  p <span style="color:#75715e">#看整合效果</span>
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;umap-by-orig.ident-after-harmony.pdf&#39;</span>,plot <span style="color:#f92672">=</span> p)
</span></span></code></pre></div><p><img alt="umap-harmony" src="/blog/post/immunead/post/immunead/umap-harmony.png"></p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>input_sce<span style="color:#f92672">=</span>seuratObj
</span></span><span style="display:flex;"><span>  input_sce <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">FindNeighbors</span>(input_sce, reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;harmony&#34;</span>,
</span></span><span style="display:flex;"><span>                             dims <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#ae81ff">15</span>) 
</span></span><span style="display:flex;"><span>  input_sce.all<span style="color:#f92672">=</span>input_sce
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  <span style="color:#75715e">#设置不同的分辨率，观察分群效果(选择哪一个？)</span>
</span></span><span style="display:flex;"><span>  <span style="color:#66d9ef">for</span> (res <span style="color:#66d9ef">in</span> <span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">0.01</span>, <span style="color:#ae81ff">0.05</span>, <span style="color:#ae81ff">0.1</span>, <span style="color:#ae81ff">0.2</span>, <span style="color:#ae81ff">0.3</span>, <span style="color:#ae81ff">0.5</span>,<span style="color:#ae81ff">0.8</span>,<span style="color:#ae81ff">1</span>)) {
</span></span><span style="display:flex;"><span>    input_sce.all<span style="color:#f92672">=</span><span style="color:#a6e22e">FindClusters</span>(input_sce.all, <span style="color:#75715e">#graph.name = &#34;CCA_snn&#34;, </span>
</span></span><span style="display:flex;"><span>                               resolution <span style="color:#f92672">=</span> res, algorithm <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>)
</span></span><span style="display:flex;"><span>  }
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">colnames</span>(input_sce.all<span style="color:#f92672">@</span>meta.data)
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">apply</span>(input_sce.all<span style="color:#f92672">@</span>meta.data[,<span style="color:#a6e22e">grep</span>(<span style="color:#e6db74">&#34;RNA_snn&#34;</span>,<span style="color:#a6e22e">colnames</span>(input_sce.all<span style="color:#f92672">@</span>meta.data))],<span style="color:#ae81ff">2</span>,table)
</span></span></code></pre></div><p><img alt="grouping" src="/blog/post/immunead/post/immunead/grouping.png"></p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p1_dim<span style="color:#f92672">=</span><span style="color:#a6e22e">plot_grid</span>(ncol <span style="color:#f92672">=</span> <span style="color:#ae81ff">3</span>, <span style="color:#a6e22e">DimPlot</span>(input_sce.all, reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;umap&#34;</span>, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;RNA_snn_res.0.01&#34;</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>                     <span style="color:#a6e22e">ggtitle</span>(<span style="color:#e6db74">&#34;louvain_0.01&#34;</span>), <span style="color:#a6e22e">DimPlot</span>(input_sce.all, reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;umap&#34;</span>, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;RNA_snn_res.0.1&#34;</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>                     <span style="color:#a6e22e">ggtitle</span>(<span style="color:#e6db74">&#34;louvain_0.1&#34;</span>), <span style="color:#a6e22e">DimPlot</span>(input_sce.all, reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;umap&#34;</span>, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;RNA_snn_res.0.2&#34;</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>                     <span style="color:#a6e22e">ggtitle</span>(<span style="color:#e6db74">&#34;louvain_0.2&#34;</span>))
</span></span><span style="display:flex;"><span>  p1_dim
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(plot<span style="color:#f92672">=</span>p1_dim, filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;Dimplot_diff_resolution_low.pdf&#34;</span>,width <span style="color:#f92672">=</span> <span style="color:#ae81ff">14</span>)
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  p1_dim<span style="color:#f92672">=</span><span style="color:#a6e22e">plot_grid</span>(ncol <span style="color:#f92672">=</span> <span style="color:#ae81ff">3</span>, <span style="color:#a6e22e">DimPlot</span>(input_sce.all, reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;umap&#34;</span>, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;RNA_snn_res.0.8&#34;</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>                     <span style="color:#a6e22e">ggtitle</span>(<span style="color:#e6db74">&#34;louvain_0.8&#34;</span>), <span style="color:#a6e22e">DimPlot</span>(input_sce.all, reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;umap&#34;</span>, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;RNA_snn_res.1&#34;</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>                     <span style="color:#a6e22e">ggtitle</span>(<span style="color:#e6db74">&#34;louvain_1&#34;</span>), <span style="color:#a6e22e">DimPlot</span>(input_sce.all, reduction <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;umap&#34;</span>, group.by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;RNA_snn_res.0.3&#34;</span>) <span style="color:#f92672">+</span> 
</span></span><span style="display:flex;"><span>                     <span style="color:#a6e22e">ggtitle</span>(<span style="color:#e6db74">&#34;louvain_0.3&#34;</span>))
</span></span><span style="display:flex;"><span>  p1_dim
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(plot<span style="color:#f92672">=</span>p1_dim, filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;Dimplot_diff_resolution_high.png&#34;</span>,width <span style="color:#f92672">=</span> <span style="color:#ae81ff">14</span>)
</span></span></code></pre></div><p><img alt="Dimplot_diff_resolution_low" src="/blog/post/immunead/post/immunead/Dimplot_diff_resolution_low.png">
<img alt="Dimplot_diff_resolution_high" src="/blog/post/immunead/post/immunead/Dimplot_diff_resolution_high.png"></p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p2_tree<span style="color:#f92672">=</span><span style="color:#a6e22e">clustree</span>(input_sce.all<span style="color:#f92672">@</span>meta.data, prefix <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;RNA_snn_res.&#34;</span>)
</span></span><span style="display:flex;"><span>  p2_tree
</span></span><span style="display:flex;"><span>  <span style="color:#a6e22e">ggsave</span>(plot<span style="color:#f92672">=</span>p2_tree, filename<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;Tree_diff_resolution.pdf&#34;</span>)
</span></span></code></pre></div><p><img alt="Tree_diff_resolution" src="/blog/post/immunead/post/immunead/Tree_diff_resolution.png"></p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#a6e22e">table</span>(input_sce.all<span style="color:#f92672">@</span>active.ident)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>  <span style="color:#ae81ff">0</span>    <span style="color:#ae81ff">1</span>    <span style="color:#ae81ff">2</span>    <span style="color:#ae81ff">3</span>    <span style="color:#ae81ff">4</span>    <span style="color:#ae81ff">5</span>    <span style="color:#ae81ff">6</span>    <span style="color:#ae81ff">7</span>    <span style="color:#ae81ff">8</span>    <span style="color:#ae81ff">9</span>   <span style="color:#ae81ff">10</span>   <span style="color:#ae81ff">11</span>   <span style="color:#ae81ff">12</span>   <span style="color:#ae81ff">13</span>   <span style="color:#ae81ff">14</span>   <span style="color:#ae81ff">15</span> 
</span></span><span style="display:flex;"><span><span style="color:#ae81ff">4885</span> <span style="color:#ae81ff">3868</span> <span style="color:#ae81ff">3818</span> <span style="color:#ae81ff">3408</span> <span style="color:#ae81ff">3223</span> <span style="color:#ae81ff">3043</span> <span style="color:#ae81ff">2941</span> <span style="color:#ae81ff">2113</span> <span style="color:#ae81ff">2103</span> <span style="color:#ae81ff">1929</span> <span style="color:#ae81ff">1817</span>  <span style="color:#ae81ff">813</span>  <span style="color:#ae81ff">769</span>  <span style="color:#ae81ff">532</span>  <span style="color:#ae81ff">280</span>  <span style="color:#ae81ff">239</span> 
</span></span><span style="display:flex;"><span>  <span style="color:#ae81ff">16</span>   <span style="color:#ae81ff">17</span> 
</span></span><span style="display:flex;"><span> <span style="color:#ae81ff">234</span>  <span style="color:#ae81ff">151</span> 
</span></span></code></pre></div>
    </div>

    
    



    
    


    <footer class="post-footer">
      <div class="post-tags">
          <a href="http://localhost:45315/blog/tags/alzheimers-disease/">Alzheimer’s disease</a>
            <a href="http://localhost:45315/blog/tags/immune-cell-signatures/">Immune Cell Signatures</a>
            <a href="http://localhost:45315/blog/tags/r/">R</a>
            <a href="http://localhost:45315/blog/tags/scrnaseq/">scRNAseq</a>
            
        </div>


      
      <nav class="post-nav">
        
          <a class="prev" href="/blog/post/pca/">
            
            <i class="iconfont">
              <svg  class="icon" viewBox="0 0 1024 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="18" height="18">
  <path d="M691.908486 949.511495l75.369571-89.491197c10.963703-12.998035 10.285251-32.864502-1.499144-44.378743L479.499795 515.267417 757.434875 204.940602c11.338233-12.190647 11.035334-32.285311-0.638543-44.850487l-80.46666-86.564541c-11.680017-12.583596-30.356378-12.893658-41.662889-0.716314L257.233596 494.235404c-11.332093 12.183484-11.041474 32.266891 0.657986 44.844348l80.46666 86.564541c1.772366 1.910513 3.706415 3.533476 5.750981 4.877077l306.620399 321.703933C662.505829 963.726242 680.945807 962.528973 691.908486 949.511495z"></path>
</svg>

            </i>
            <span class="prev-text nav-default">Principle component analysis (PCA)</span>
            <span class="prev-text nav-mobile">Prev</span>
          </a>
        
          <a class="next" href="/blog/post/osteoarthrisis/">
            <span class="next-text nav-default">A osteoarthritis study using scRNAseq data</span>
            <span class="prev-text nav-mobile">Next</span>
            
            <i class="iconfont">
              <svg class="icon" viewBox="0 0 1024 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="18" height="18">
  <path d="M332.091514 74.487481l-75.369571 89.491197c-10.963703 12.998035-10.285251 32.864502 1.499144 44.378743l286.278095 300.375162L266.565125 819.058374c-11.338233 12.190647-11.035334 32.285311 0.638543 44.850487l80.46666 86.564541c11.680017 12.583596 30.356378 12.893658 41.662889 0.716314l377.434212-421.426145c11.332093-12.183484 11.041474-32.266891-0.657986-44.844348l-80.46666-86.564541c-1.772366-1.910513-3.706415-3.533476-5.750981-4.877077L373.270379 71.774697C361.493148 60.273758 343.054193 61.470003 332.091514 74.487481z"></path>
</svg>

            </i>
          </a>
      </nav>
    </footer>
  </article>

  
  


  
  

        </div>
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="icon-links">
  
  
    <a href="mailto:lyqglyqg@gmail.com" rel="me noopener" class="iconfont"
      title="email" >
      <svg class="icon" viewBox="0 0 1451 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="36" height="36">
  <path d="M664.781909 681.472759 0 97.881301C0 3.997201 71.046997 0 71.046997 0L474.477909 0 961.649408 0 1361.641813 0C1361.641813 0 1432.688811 3.997201 1432.688811 97.881301L771.345323 681.472759C771.345323 681.472759 764.482731 685.154773 753.594283 688.65053L753.594283 688.664858C741.602731 693.493018 729.424896 695.068979 718.077952 694.839748 706.731093 695.068979 694.553173 693.493018 682.561621 688.664858L682.561621 688.65053C671.644501 685.140446 664.781909 681.472759 664.781909 681.472759L664.781909 681.472759ZM718.063616 811.603883C693.779541 811.016482 658.879232 802.205449 619.10784 767.734955 542.989056 701.759633 0 212.052267 0 212.052267L0 942.809523C0 942.809523 0 1024 83.726336 1024L682.532949 1024 753.579947 1024 1348.948139 1024C1432.688811 1024 1432.688811 942.809523 1432.688811 942.809523L1432.688811 212.052267C1432.688811 212.052267 893.138176 701.759633 817.019477 767.734955 777.248 802.205449 742.347691 811.03081 718.063616 811.603883L718.063616 811.603883Z"></path>
</svg>

    </a>
  
    <a href="http://lyqg.github.io/" rel="me noopener" class="iconfont"
      title="github"  target="_blank"
      >
      <svg class="icon" style="" viewBox="0 0 1024 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="36" height="36">
  <path d="M512 12.672c-282.88 0-512 229.248-512 512 0 226.261333 146.688 418.133333 350.08 485.76 25.6 4.821333 34.986667-11.008 34.986667-24.618667 0-12.16-0.426667-44.373333-0.64-87.04-142.421333 30.890667-172.458667-68.693333-172.458667-68.693333C188.672 770.986667 155.008 755.2 155.008 755.2c-46.378667-31.744 3.584-31.104 3.584-31.104 51.413333 3.584 78.421333 52.736 78.421333 52.736 45.653333 78.293333 119.850667 55.68 149.12 42.581333 4.608-33.109333 17.792-55.68 32.426667-68.48-113.706667-12.8-233.216-56.832-233.216-253.013333 0-55.893333 19.84-101.546667 52.693333-137.386667-5.76-12.928-23.04-64.981333 4.48-135.509333 0 0 42.88-13.738667 140.8 52.48 40.96-11.392 84.48-17.024 128-17.28 43.52 0.256 87.04 5.888 128 17.28 97.28-66.218667 140.16-52.48 140.16-52.48 27.52 70.528 10.24 122.581333 5.12 135.509333 32.64 35.84 52.48 81.493333 52.48 137.386667 0 196.693333-119.68 240-233.6 252.586667 17.92 15.36 34.56 46.762667 34.56 94.72 0 68.522667-0.64 123.562667-0.64 140.202666 0 13.44 8.96 29.44 35.2 24.32C877.44 942.592 1024 750.592 1024 524.672c0-282.752-229.248-512-512-512"></path>
</svg>

    </a>


<a href="http://localhost:45315/blog/index.xml" rel="noopener alternate" type="application/rss&#43;xml"
    class="iconfont" title="rss" target="_blank">
    <svg class="icon" viewBox="0 0 1024 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="30" height="30">
  <path d="M819.157333 1024C819.157333 574.592 449.408 204.8 0 204.8V0c561.706667 0 1024 462.293333 1024 1024h-204.842667zM140.416 743.04a140.8 140.8 0 0 1 140.501333 140.586667A140.928 140.928 0 0 1 140.074667 1024C62.72 1024 0 961.109333 0 883.626667s62.933333-140.544 140.416-140.586667zM678.784 1024h-199.04c0-263.210667-216.533333-479.786667-479.744-479.786667V345.173333c372.352 0 678.784 306.517333 678.784 678.826667z"></path>
</svg>

  </a>
  
</div>

<div class="copyright">
  <span class="power-by">
    Powered by <a class="hexo-link" href="https://gohugo.io">Hugo</a>
  </span>
  <span class="division">|</span>
  <span class="theme-info">
    Theme - <a class="theme-link" href="https://github.com/xianmin/hugo-theme-jane">Jane</a>
  </span>

  <span class="copyright-year">
    &copy;
    
      2023 -
    2024
    <span class="heart">
      
      <i class="iconfont">
        <svg class="icon" viewBox="0 0 1025 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="14" height="14">
  <path d="M1000.1 247.9c-15.5-37.3-37.6-70.6-65.7-98.9-54.4-54.8-125.8-85-201-85-85.7 0-166 39-221.4 107.4C456.6 103 376.3 64 290.6 64c-75.1 0-146.5 30.4-201.1 85.6-28.2 28.5-50.4 61.9-65.8 99.3-16 38.8-24 79.9-23.6 122.2 0.7 91.7 40.1 177.2 108.1 234.8 3.1 2.6 6 5.1 8.9 7.8 14.9 13.4 58 52.8 112.6 102.7 93.5 85.5 209.9 191.9 257.5 234.2 7 6.1 15.8 9.5 24.9 9.5 9.2 0 18.1-3.4 24.9-9.5 34.5-30.7 105.8-95.9 181.4-165 74.2-67.8 150.9-138 195.8-178.2 69.5-57.9 109.6-144.4 109.9-237.3 0.1-42.5-8-83.6-24-122.2z"
   fill="#8a8a8a"></path>
</svg>

      </i>
    </span><span class="author">
        lyqg
        
      </span></span>

  
  

  
</div>

    </footer>

    <div class="back-to-top" id="back-to-top">
      <i class="iconfont">
        
        <svg class="icon" viewBox="0 0 1024 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="35" height="35">
  <path d="M510.866688 227.694839 95.449397 629.218702l235.761562 0-2.057869 328.796468 362.40389 0L691.55698 628.188232l241.942331-3.089361L510.866688 227.694839zM63.840492 63.962777l894.052392 0 0 131.813095L63.840492 195.775872 63.840492 63.962777 63.840492 63.962777zM63.840492 63.962777"></path>
</svg>

      </i>
    </div>
  </div>
  
<script type="text/javascript" src="/blog/lib/jquery/jquery-3.2.1.min.js"></script>
  <script type="text/javascript" src="/blog/lib/slideout/slideout-1.0.1.min.js"></script>




<script type="text/javascript" src="/blog/js/main.fe83e11b4fbc9193d67e2c9db78bad21f8dc59fca0cacd8c1c3bb071bb16a852.js" integrity="sha256-/oPhG0&#43;8kZPWfiydt4utIfjcWfygys2MHDuwcbsWqFI=" crossorigin="anonymous"></script>












  
    <script type="text/javascript" src="/js/load-photoswipe.js"></script>
    <script type="text/javascript" src="/lib/photoswipe/photoswipe.min.js"></script>
    <script type="text/javascript" src="/lib/photoswipe/photoswipe-ui-default.min.js"></script>
  

















</body>
</html>
